core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java

```diff
@@ -0,0 +1,180 @@
+                }
+                break;
+            case LEFT:
+                if (lefSourceSingleConstantRow && isAtLeastScalar(right, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+                break;
+            case RIGHT:
+                if (lefSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow && isAtLeastScalar(left, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+                break;
+            case FULL:
+                if (lefSourceSingleConstantRow && isAtLeastScalar(right, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow && isAtLeastScalar(left, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+        }
+
+        return Result.empty();
+    }
+
+    private static boolean isUnconditional(JoinNode joinNode)
+    {
+        return joinNode.getCriteria().isEmpty() &&
+                (joinNode.getFilter().isEmpty() || joinNode.getFilter().get().equals(TRUE_LITERAL));
+    }
+
+    private boolean isSingleConstantRow(PlanNode node)
+    {
+        if (!(node instanceof ValuesNode)) {
+            return false;
+        }
+
+        ValuesNode values = (ValuesNode) node;
+        if (values.getRowCount() != 1) {
+            return false;
+        }
+
+        if (values.getOutputSymbols().isEmpty()) {
+            // handled by ReplaceRedundantJoinWithSource rule
+            return false;
+        }
+
+        Expression row = getOnlyElement(values.getRows().get());
+
+        return row instanceof Row;
+    }
+
+    private ProjectNode appendProjection(PlanNode source, List<Symbol> sourceOutputs, PlanNode constantSource, List<Symbol> constantOutputs, PlanNodeIdAllocator idAllocator)
+    {
+        Assignments.Builder assignments = Assignments.builder()
+                .putIdentities(sourceOutputs);
```

Comment:
Move this down closer to where it's used. I would restructure this method as follows for better visual clarity:

```
        ValuesNode values = (ValuesNode) constantSource;
        Row row = (Row) getOnlyElement(values.getRows().get());

        Map<Symbol, Expression> mapping = new HashMap<>();
        for (int i = 0; i < values.getOutputSymbols().size(); i++) {
            mapping.put(values.getOutputSymbols().get(i), row.getItems().get(i));
        }

        Assignments.Builder assignments = Assignments.builder()
                .putIdentities(sourceOutputs);

        constantOutputs.stream()
                .forEach(symbol -> assignments.put(symbol, mapping.get(symbol)));

        return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
```